lib/cocoapods/installer/project_cache/project_cache_analyzer.rb

```diff
@@ -86,38 +75,81 @@ def analyze
           changed_pod_targets, changed_aggregate_targets = changed_targets.partition { |target| target.is_a?(PodTarget) }
 
-          pod_targets_to_generate = changed_pod_targets + added_pod_targets
+          dirty_targets = compute_dirtied_targets(pod_targets + aggregate_targets)
+          dirty_pod_targets, dirty_aggregate_targets = dirty_targets.partition { |target| target.is_a?(PodTarget) }
+
+          pod_targets_to_generate = changed_pod_targets + added_pod_targets + dirty_pod_targets
 
           # We either return the full list of aggregate targets or none since the aggregate targets go into the Pods.xcodeproj
           # and so we need to regenerate all aggregate targets when regenerating Pods.xcodeproj.
           aggregate_target_to_generate =
-            unless (changed_aggregate_targets + added_aggregate_targets + removed_aggregate_targets).empty?
+            unless (changed_aggregate_targets + added_aggregate_targets + removed_aggregate_targets + dirty_aggregate_targets).empty?
               aggregate_targets
             end
 
           ProjectCacheAnalysisResult.new(pod_targets_to_generate, aggregate_target_to_generate, cache_key_by_target_label,
                                          build_configurations, project_object_version)
         end
+
+        private
+
+        def create_cache_key_mappings(target_by_label)
+          Hash[target_by_label.map do |label, target|
+            case target
+            when PodTarget
+              local = sandbox.local?(target.pod_name)
+              checkout_options = sandbox.checkout_sources[target.pod_name]
+              [label, TargetCacheKey.from_pod_target(target, :is_local_pod => local, :checkout_options => checkout_options)]
+            when AggregateTarget
+              [label, TargetCacheKey.from_aggregate_target(target)]
+            else
+              raise "[BUG] Unknown target type #{target}"
+            end
+          end]
+        end
+
+        def compute_added_and_removed_targets(target_by_label, target_labels, cached_target_labels)
+          added_targets = (target_labels - cached_target_labels).map do |label|
+            target_by_label[label]
+          end
+          removed_targets = (cached_target_labels - target_labels).map do |label|
+            target_by_label[label]
+          end
+          [added_targets, removed_targets]
+        end
+
+        def compute_changed_targets_from_cache(cache_key_by_target_label, target_by_label, cache)
+          cache_key_by_target_label.each_with_object([]) do |(label, cache_key), changed_targets|
+            next unless cache.cache_key_by_target_label[label]
+            if cache_key.key_difference(cache.cache_key_by_target_label[label]) == :project
+              changed_targets << target_by_label[label]
+            end
+          end
+        end
+
+        def compute_dirtied_targets(targets)
```

Comment:
`compute_dirty_targets`?

Reply:
meh I guess its fine

Reply:
Nah i think your suggestion is better. Will fix